The entire code base of the plugin

amwolff opened the pull request:
cc @iomodo

iomodo commented on server/plugin.go:
```diff
@@ -1,28 +1,117 @@
+	return nil
+}
+
+var defaultKVStore = map[string]interface{}{
```
why not have these in the plugin settings?

amwolff replied:
Done.

iomodo commented on server/report.go:
```diff
@@ -0,0 +1,397 @@
+
+var reportLock = []byte{0b10, 0b1, 0b11, 0b111}
+
+func (p *Plugin) monitoringReportLoop() {
```
In the HA environment this method will be run on every single node. Do we want that?

amwolff replied:
Yes. I wanted to ensure the system monitoring report's continuous delivery even if one of the plugin instances die (in the HA environment). Since it's not enough to run the reporting loop on one node and running the same code on every node will result in too many reports delivered, I have implemented mutual exclusion (using KV store) of report delivery in `monitoringReportLoop`. All nodes launch this function, but only one node enters the so-called critical section.

iomodo commented on server/report.go:
```diff
@@ -0,0 +1,397 @@
+var reportLock = []byte{0b10, 0b1, 0b11, 0b111}
+
+func (p *Plugin) monitoringReportLoop() {
+	for {
```
This looks like a busy waiting. why not use a cron job instead?

amwolff replied:
I believe that's not enough for the HA environment. If I understand your suggestion correctly, the _cron job_ would be running the reporting loop on the node selected after subscription. I'm afraid that when the plugin instance on that node dies, the system monitoring report will be no longer delivered. Instead, I'm proposing a solution where every node runs the loop, but only one of them enters the critical section. Every node waits a minute (using `time.Sleep`) and tries to enter the critical section to build the report. The lock has a duration of the system monitoring report frequency. Only one node does that since this is a mutually exclusive action. I also believe this is approximately as efficient as a non-HA solution; there's no busy waiting in its classical definition because we use `time.Sleep`.

iomodo replied:
I'd encourage using the scheduling mechanism from [here](https://github.com/mattermost/mattermost-plugin-api/blob/1d382e5ea2077dfd81933aed74222fd5b7e2f8cf/cluster/job.go#L110). It is implemented using the distributed lock and you will be guaranteed to have one `monitoringReportLoop` run in every X seconds in an HA environment.

amwolff replied:
Done. Thank you. It made this part of the code substantially leaner.

iomodo commented on server/report.go:
```diff
@@ -0,0 +1,397 @@
+	return b.String()
+}
+
+type extractedService struct {
```
`extractedService` and `extractedHost` are the same struct why do we need both?

amwolff replied:
Thanks for noticing that. These structs have probably differed initially but eventually became the same. I have merged them into one now. :)

iomodo commented on README.md:
```diff
@@ -1,117 +1,250 @@
+    2. Click *Regenerate* to regenerate the token
+    3. Copy the token (you are going to use it later)
+2. Click *Save* to save the settings
+3. Switch to the machine where Nagios is running (preferably)
```
This looks invasive, does nagios API provide similar functionality?

amwolff replied:
Unfortunately not. :( The Nagios API does not provide such functionality whatsoever. It's even worse: Nagios have no idea its configuration has changed while it's running. The only way to reload the configuration is to restart it.

I agree this looks invasive. I made this point on the Plugin: Nagios channel ([link to the discussion](https://community.mattermost.com/core/pl/r3rgadinzjf4xdqpgsjzu87xah)), and we have decided it's the best way to do this monitoring thing right now.

iomodo commented on server/report.go:
```diff
@@ -0,0 +1,397 @@
+package main
```
Can we split up these big files into separate ones? That would be easier to maintain I think.

amwolff replied:
Done.

iomodo reviewed (changes requested):
Great work!
Have couple of questions below. I have not reviewed the `watcher` part yet. Not sure if watching the actual file system approach is something we want to do. Main question there would be can this part be changed with the Nagios api?

Also would be great to have more unit tests.

Thanks for the submission! Looks really promising!

amwolff commented on server/command.go:
```diff
@@ -0,0 +1,235 @@
+	return command, action, parameters
+}
+
+func (p *Plugin) getCommandResponse(
```
Done.

amwolff commented on server/emoji.go:
```diff
@@ -0,0 +1,47 @@
+	warningEmoji              = ":warning:"
+)
+
+func emoji(state string) string {
```
Done.

amwolff commented on server/plugin.go:
```diff
@@ -1,28 +1,117 @@
+}
+
+func (p *Plugin) OnActivate() error {
+	botUserID, err := p.Helpers.EnsureBot(&model.Bot{
```
I'm not sure whether this won't break anything, so I'm skipping this suggestion now (if that's ok). When writing code for creating the bot and other things, I have mostly followed conventions set in other plugins.

amwolff commented:
@iomodo Thanks a lot for the review.

A brief description of the changes I have made since your review:

- Renamed confusing `getCommandResponse`
- Made a map to look up the emojis
- Moved default settings to the plugin configuration
- Made the system monitoring report a thread
- Made `extractedHost` and `extractedService` one struct
- Split `report.go`
- Added more unit tests

For these changes I haven't made, I tried to provide sufficient explanation in the comments above. Using Nagios API to monitor the configuration changes is one of them. Unfortunately, it's not possible with Nagios.

Let me know what you think! :)

iomodo reviewed (changes requested):
Just one thing about the HA and the scheduled job. Other than that it looks good to me.

amwolff commented:
@iomodo I have rewritten the monitoring report loop using `cluster.Schedule`. Thank you for the helpful suggestions. :shipit:

iomodo reviewed (dismissed):
LGTM, @amwolff thanks for the changes!

levb reviewed (changes requested):
1. in the high availability mode (HA) `monitoringReportLoop` will run on every host, resulting in message duplication, no? Please see https://github.com/mattermost/mattermost-plugin-api/tree/master/cluster for an example of setting up a "job" in a plugin.
2. I am a little confused as to why the Report ChannelID and Configuration ChannelID are saved in the KVStore and not in the config. The KV store is eventually consistent, so in theory the values might not be available to read right away; storing these values in the `config` and using `API.StoreConfig` seems a more robust approach. I recommend changing to store these values in config; but it will probably work well enough as is.

amwolff commented:
Hi @levb. Thanks a lot for the review. :)

1. I can see that you might have reviewed the initial version of this PR. In the current version (https://github.com/ulumuri/mattermost-plugin-nagios/pull/16/commits/31a23d6266ea1e035ff891073f5863fd7b717f9a), the system monitoring report mechanism already uses `cluster`'s instrumentation to schedule the report as a "job". Although I don't think the initial version suffers from message duplication (I have implemented a locking mechanism based on KVStore), I can see that the way I did it triggered some confusion, so I'm grateful that @iomodo and you have pointed that out;
2. I must admit I had no idea that KVStore is eventually consistent. Maybe we should point this out [here](https://developers.mattermost.com/extend/plugins/server/ha/). I'm also wondering which state-related values we should keep in KVStore and which we shouldn't. I believe that most (if not all) of these values (like Report Channel ID) will be accessed, on average, as frequently as others. Should I keep them all in the config? What advantages does KVStore have then?

catalintomai commented:
@levb, @iomodo - are we ok on the message duplication issue then?

amwolff commented:
I will now change the base to `master`, merge the changes, make a new release, and transfer the repository.
